lib/docker.js

```diff
@@ -135,24 +140,50 @@ function createDockerFile() {
 
     // set the deploy target
     baseImg = targets[pkg.deploy.target];
+    if (!baseImg || baseImg === '') {
+        console.error('ERROR: You must specify a valid target!');
+        console.error('ERROR: Check the deploy stanza in package.json and targets.yaml');
+        process.exit(2);
+    }
+
     // get any additional packages that need to be installed
     Object.keys(pkg.deploy.dependencies).forEach(function(sys) {
-        if (sys !== '_all' && (sys === baseImg || (new RegExp(sys)).test(baseImg))) {
-            Array.prototype.push.apply(extraPkgs, pkg.deploy.dependencies[sys]);
+        if (sys === '_all' || (sys === baseImg || (new RegExp(sys)).test(baseImg))) {
+            pkg.deploy.dependencies[sys].forEach(function(pkg) {
+                if (typeof pkg === 'string') {
+                    extraPkgs.push(pkg);
+                } else if (typeof pkg === 'object') {
+                    if (!pkg.repo_url || !pkg.pool || !pkg.packages || !pkg.release) {
+                        console.error('ERROR: Incorrect dependency spec: ' + JSON.stringify(pkg));
+                    }
+                    customSourcePkgs.push(pkg);
+                } else {
+                    console.error('ERROR: Incorrect dependency spec: ' + pkg);
+                    process.exit(2);
+                }
+            });
         }
     });
-    Array.prototype.push.apply(extraPkgs, pkg.deploy.dependencies._all);
 
-    if (!baseImg || baseImg === '') {
-        console.error('ERROR: You must specify a valid target!');
-        console.error('ERROR: Check the deploy stanza in package.json and targets.yaml');
-        process.exit(2);
+    if (customSourcePkgs.some(function(customSourcePkgSpec) {
+            return /^https/.test(customSourcePkgSpec.repo_url);
+        })) {
+        extraPkgs.push('apt-transport-https');
     }
 
     contents = 'FROM ' + baseImg + '\n' +
-        'RUN apt-get update --fix-missing && apt-get install -y ' + extraPkgs.join(' ') +
+        'RUN apt-get update && apt-get install -y ' + extraPkgs.join(' ') +
         ' && rm -rf /var/lib/apt/lists/*\n';
 
+    customSourcePkgs.forEach(function(customSourcePkgSpec) {
+        contents += 'RUN echo "deb ' + customSourcePkgSpec.repo_url +
```

Comment:
No no, I was thinking just of making an optimisation here: if there are multiple sources, add them all first, then update and install packages from these spurces. The packages from the main pool should always be installed before that precisely to avoid the problems you are describing.